engine/crates/grafbase-engine/src/resolver_utils/list.rs

```diff
@@ -1,129 +1,201 @@
+        MetaFieldType, MetaType, WrappingType, WrappingTypeIter,
     },
     resolver_utils::resolve_container,
     ContextSelectionSet, Error, LegacyOutputType, Positioned, ServerError, ServerResult, Value,
 };
 
-/// Resolve an list by executing each of the items concurrently.
+/// Resolve a list by executing each of the items concurrently.
 pub async fn resolve_list<'a>(
-    ctx: &ContextSelectionSet<'a>,
+    ctx: ContextSelectionSet<'a>,
     field: &Positioned<Field>,
-    ty: &'a MetaType,
-    values: Vec<ResolvedValue>,
+    field_ty: &MetaFieldType,
+    inner_ty: &'a MetaType,
+    value: ResolvedValue,
 ) -> ServerResult<ResponseNodeId> {
-    let extensions = &ctx.query_env.extensions;
-    if !extensions.is_empty() {
-        let mut futures = Vec::with_capacity(values.len());
-        for (idx, item) in values.into_iter().enumerate() {
-            futures.push({
-                let ctx = ctx.clone();
-                let type_name = ty.name();
-                async move {
-                    let ctx_idx = ctx.with_index(idx, Some(&ctx.item.node));
-                    let extensions = &ctx.query_env.extensions;
+    #[async_recursion::async_recursion]
+    async fn inner(
+        ctx: ContextSelectionSet<'async_recursion>,
+        field: &Positioned<Field>,
+        list_kinds: &[ListKind],
+        ty: &MetaType,
+        value: ResolvedValue,
+    ) -> Result<ResponseNodeId, ServerError> {
+        let Some(list_kind) = list_kinds.first() else {
+            // If there's no list_kind then we've reached the innermost item and should resolve that
+            return resolve_item(ctx, field, ty, value).await;
+        };
 
-                    let ctx_field = ctx.with_field(field, None, Some(&ctx.item.node));
-                    let meta_field = ctx_field
-                        .schema_env
-                        .registry
-                        .types
-                        .get(type_name)
-                        .and_then(|ty| ty.field_by_name(field.node.name.node.as_str()));
+        // First we need to make sure our parent resolve data actually has a list
+        // (or return null early if we're on a nullable list)
+        let items = match (list_kind, value.data_resolved()) {
+            (ListKind::NullableList(_), serde_json::Value::Null) => {
+                return Ok(ctx.response_graph.write().await.insert_node(CompactValue::Null));
+            }
+            (ListKind::NonNullList(_), serde_json::Value::Null) => {
+                return Err(ctx.set_error_path(ServerError::new(
+                    format!(
+                        "An error occurred while fetching `{}`, a non-nullable value was expected but no value was found.",
+                        field.node.name.node
+                    ),
+                    Some(ctx.item.pos),
+                )));
+            }
+            (_, serde_json::Value::Array(_)) => value.item_iter().expect("we checked its an array").collect::<Vec<_>>(),
```

Comment:
might be worth a comment that you're using `item_ter` to avoid a copy of the data.